test/activerecord_spanner_mock_server/read_only_transaction_test.rb

```diff
@@ -111,9 +111,67 @@ def test_multiple_snapshots
-      @mock.put_statement_result sql, MockServerTests::create_random_singers_result(1)
+    def test_single_max_staleness
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("max_staleness: 5.45").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.max_staleness
+      assert_equal 5, single_use.read_only.max_staleness.seconds
+      assert_equal 450000000, single_use.read_only.max_staleness.nanos
+    end
+
+    def test_single_exact_staleness
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("exact_staleness:100").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.exact_staleness
+      assert_equal 100, single_use.read_only.exact_staleness.seconds
+      assert_equal 0, single_use.read_only.exact_staleness.nanos
+    end
+
+    def test_single_min_read_timestamp
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("min_read_timestamp: 2021-09-07T14:33:30.1123Z").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.min_read_timestamp
+      time = Google::Cloud::Spanner::Convert.time_to_timestamp Time.xmlschema("2021-09-07T14:33:30.1123Z")
+      assert_equal time, single_use.read_only.min_read_timestamp
+    end
+
+    def test_single_read_timestamp
+      sql = register_singer_find_by_id_result true
+
+      Singer.optimizer_hints("read_timestamp: 2021-09-07T14:33:31Z").find_by(id: 1)
+
+      execute_sql_request = @mock.requests.find { |req| req.is_a?(Google::Cloud::Spanner::V1::ExecuteSqlRequest) && req.sql == sql }
+      single_use = execute_sql_request.transaction.single_use
+      assert single_use
+      assert single_use.read_only.return_read_timestamp
+      assert single_use.read_only.read_timestamp
+      time = Google::Cloud::Spanner::Convert.time_to_timestamp Time.xmlschema("2021-09-07T14:33:31.000Z")
+      assert_equal time, single_use.read_only.read_timestamp
+    end
+
+    def register_singer_find_by_id_result with_hints = false
+      res = MockServerTests::create_random_singers_result(1)
+      sql = "SELECT  `singers`.* FROM `singers` WHERE `singers`.`id` = @p1 LIMIT @p2" if with_hints
```

Comment:
nits: two spaces after SELECT.

Reply:
what are the difference between these two sql strings?

Reply:
The reason that both strings are added is that ActiveRecord will always generate a SQL string with an additional space if there is at least one optimizer hint. Normally an optimizer hint is added as a comment in the SQL string, which is why ActiveRecord adds an extra space (to add the space after the comment), but as we translate at least some of the hints into transaction options, this just leads to a SQL string with an extra space. That part of the SQL generation in ActiveRecord is not really easy to override.

In real life the extra space does not make any difference at all, but the mock server that we are using for testing is very simple. So it sees the SQL string with an extra space as a different query than the one without it, which is why we need both.